// @uniswap/universal-router/typechain/factories/BytesLib__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { BytesLib } from "../BytesLib";

export class BytesLib__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<BytesLib> {
    return super.deploy(overrides || {}) as Promise<BytesLib>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): BytesLib {
    return super.attach(address) as BytesLib;
  }
  connect(signer: Signer): BytesLib__factory {
    return super.connect(signer) as BytesLib__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): BytesLib {
    return new Contract(address, _abi, signerOrProvider) as BytesLib;
  }
}

const _abi = [
  {
    inputs: [],
    name: "NoSlice",
    type: "error",
  },
  {
    inputs: [],
    name: "SliceOutOfBounds",
    type: "error",
  },
  {
    inputs: [],
    name: "SliceOverflow",
    type: "error",
  },
  {
    inputs: [],
    name: "ToAddressOutOfBounds",
    type: "error",
  },
  {
    inputs: [],
    name: "ToAddressOverflow",
    type: "error",
  },
  {
    inputs: [],
    name: "ToUint24OutOfBounds",
    type: "error",
  },
  {
    inputs: [],
    name: "ToUint24Overflow",
    type: "error",
  },
];

const _bytecode =
  "0x6080806040523460175760119081601d823930815050f35b600080fdfe600080fdfea164736f6c6343000811000a";
