// @uniswap/universal-router/typechain/factories/ICryptoPunksMarket__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "@ethersproject/providers";

import type { ICryptoPunksMarket } from "../ICryptoPunksMarket";

export class ICryptoPunksMarket__factory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ICryptoPunksMarket {
    return new Contract(address, _abi, signerOrProvider) as ICryptoPunksMarket;
  }
}

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "punkIndex",
        type: "uint256",
      },
    ],
    name: "buyPunk",
    outputs: [],
    stateMutability: "payable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "to",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "punkIndex",
        type: "uint256",
      },
    ],
    name: "transferPunk",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];
