// @uniswap/universal-router/typechain/factories/ReentrancyLock__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { ReentrancyLock } from "../ReentrancyLock";

export class ReentrancyLock__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<ReentrancyLock> {
    return super.deploy(overrides || {}) as Promise<ReentrancyLock>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): ReentrancyLock {
    return super.attach(address) as ReentrancyLock;
  }
  connect(signer: Signer): ReentrancyLock__factory {
    return super.connect(signer) as ReentrancyLock__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ReentrancyLock {
    return new Contract(address, _abi, signerOrProvider) as ReentrancyLock;
  }
}

const _abi = [
  {
    inputs: [],
    name: "ContractLocked",
    type: "error",
  },
];

const _bytecode =
  "0x60808060405234601857600160005560119081601e8239f35b600080fdfe600080fdfea164736f6c6343000811000a";
