// @uniswap/universal-router/typechain/factories/UnsupportedProtocol__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { UnsupportedProtocol } from "../UnsupportedProtocol";

export class UnsupportedProtocol__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<UnsupportedProtocol> {
    return super.deploy(overrides || {}) as Promise<UnsupportedProtocol>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): UnsupportedProtocol {
    return super.attach(address) as UnsupportedProtocol;
  }
  connect(signer: Signer): UnsupportedProtocol__factory {
    return super.connect(signer) as UnsupportedProtocol__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): UnsupportedProtocol {
    return new Contract(address, _abi, signerOrProvider) as UnsupportedProtocol;
  }
}

const _abi = [
  {
    inputs: [],
    name: "UnsupportedProtocolError",
    type: "error",
  },
  {
    stateMutability: "nonpayable",
    type: "fallback",
  },
];

const _bytecode =
  "0x608080604052346013576043908160198239f35b600080fdfe60808060405234603157807fea3559ef0000000000000000000000000000000000000000000000000000000060049252fd5b600080fdfea164736f6c6343000811000a";
