// trpc-sveltekit/dist/server.js
import { resolveHTTPResponse } from '@trpc/server/http';
/**
 * Create a SvelteKit handle function for rRPC requests.
 *
 * If you want to use it in conjunction with other SvelteKit handles,
 * consider [the sequence helper function](https://kit.svelte.dev/docs/modules#sveltejs-kit-hooks).
 * @see https://kit.svelte.dev/docs/hooks
 */
export function createTRPCHandle({ router, url = '/trpc', createContext, responseMeta, onError }) {
    return async ({ event, resolve }) => {
        if (event.url.pathname.startsWith(url)) {
            const request = event.request;
            const req = {
                method: request.method,
                headers: request.headers,
                query: event.url.searchParams,
                body: await request.text()
            };
            const httpResponse = await resolveHTTPResponse({
                router,
                req,
                path: event.url.pathname.substring(url.length + 1),
                createContext: async () => createContext?.(event),
                responseMeta,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                onError: onError
            });
            const { status, headers, body } = httpResponse;
            return new Response(body, { status, headers });
        }
        return resolve(event);
    };
}
